// print_control/SocketControl.js
/**
 * Created by Administrator on 2015/1/23.
 */

var source = require('print_source');
var nodePlatSource = source.nodePlatSource;

var util = require('print_util');
var mongoDBUtil = util.mongoDBUtil;
var log = util.log;

var cons = require('print_constants');
var socketCons = cons.socketCons;

var assist = require('print_assist');
var queryAssist = assist.queryAssist;


var SocketControl = function () {
};

SocketControl.prototype.handle = function (cmd, bodyNode, cb) {
    var self = this;
    if (typeof self[cmd] === 'function') {
        self[cmd](bodyNode, cb);
    } else {
        console.log('找不到处理此命令的方法: ' + cmd);
        cb();
    }
};

//登录
SocketControl.prototype.login = function (bodyNode, cb) {
    var self = this;
    var userName = bodyNode.userName;
    self.userName = userName;
    console.log(userName + '已登录');
    cb(null, socketCons.toSelf, 'login', 'ok');
};

//更改取票任务状态
SocketControl.prototype.catchTicketsStatus = function (bodyNode, cb) {
    var status = nodePlatSource.catchWaitQueenTaskStatus;
    if (status) {
        nodePlatSource.catchWaitQueenTask.stop();
        nodePlatSource.catchWaitQueenTaskStatus = false;
    } else {
        nodePlatSource.catchWaitQueenTask.start();
        nodePlatSource.catchWaitQueenTaskStatus = true;
    }
    ;
    cb(null, socketCons.toAll, 'catchTicketsStatus', nodePlatSource.catchWaitQueenTaskStatus);
};

//查询取票任务状态
SocketControl.prototype.queryCatchTicketsStatus = function (bodyNode, cb) {
    cb(null, socketCons.toSelf, 'queryCatchTicketsStatus', nodePlatSource.catchWaitQueenTaskStatus);
};


//查询用户列表
SocketControl.prototype.userList = function (bodyNode, cb) {
    mongoDBUtil.db.collection('customer', {safe: true}, function (err, collection) {
        collection.find().toArray(function (err, datas) {
            cb(err, socketCons.toSelf, 'userList', datas);
        })
    });
};

SocketControl.prototype.addUser = function (bodyNode, cb) {
    mongoDBUtil.db.collection('customer', {safe: true}, function (err, collection) {
        collection.find({'userName': bodyNode.userName}).toArray(function (err, datas) {
            if (datas.length == 0) {
                collection.insert(bodyNode, function () {
                    collection.find().toArray(function (err, datas) {
                        cb(err, socketCons.toAll, 'userList', datas);
                    });
                });
            } else {
                collection.findAndModify({'userName': bodyNode.userName}, {'id': -1}, {$set: bodyNode}, {new: true}, function (err, result) {
                    collection.find().toArray(function (err, datas) {
                        cb(err, socketCons.toAll, 'userList', datas);
                    });
                });
            };
        });
    });
};

SocketControl.prototype.terminalList = function (bodyNode, cb) {
    mongoDBUtil.db.collection('terminal', {safe: true}, function (err, collection) {
        collection.find().toArray(function (err, datas) {
            cb(err, socketCons.toSelf, 'terminalList', datas);
        })
    });
};

//请求获取等待队列
SocketControl.prototype.waitQueen = function (bodyNode, cb) {
    queryAssist.getWaitQueen(function (err, datas) {
        if (!err) {
            var waitQueen = datas;
            cb(err, socketCons.toSelf, 'waitQueen', waitQueen);
        }
    });
};


SocketControl.prototype.addTerminal = function (bodyNode, cb) {
    mongoDBUtil.db.collection('terminal', {safe: true}, function (err, collection) {
        collection.find({'id': bodyNode.id}).toArray(function (err, datas) {
            if (datas.length == 0) {
                bodyNode.status = 1900;
                collection.insert(bodyNode, function () {
                    cb(err, socketCons.toAll, 'addTerminal', bodyNode);
                });
            } else {
                collection.findAndModify({'id': bodyNode.id}, {'id': -1}, {$set: bodyNode}, {new: true}, function (err, result) {
                    cb(err, socketCons.toAll, 'editTerminal', result);
                });
            };
        });
    });
};

module.exports = SocketControl;
