// print_control/SocketControl.js
/**
 * Created by Administrator on 2015/1/23.
 */
var testSource = require('../print_source/TestSource');
var db = require('../print_util/MongoDBUtil');
var cons=require('print_constants');
var socketCons=cons.socketCons;


var SocketControl = function () {};

SocketControl.prototype.handle = function(cmd, bodyNode, cb)
{
    var self = this;
    if (typeof self[cmd] === 'function') {
        self[cmd](bodyNode,cb);
    } else {
        console.log('找不到处理此命令的方法: '+cmd);
        cb();
    }
};

//登录
SocketControl.prototype.login = function (bodyNode,cb) {
    var self=this;
    var userName =  bodyNode.userName;
    self.userName = userName;
    console.log( userName+'已登录');
    cb(null,socketCons.toSelf,'login','ok');
};

//更改取票任务状态
SocketControl.prototype.catchTicketsStatus = function (bodyNode,cb) {
    var status = testSource.catchWaitQueenTaskStatus;
    if(status){
        testSource.catchWaitQueenTask.stop();
        testSource.catchWaitQueenTaskStatus=false;
    }else{
        testSource.catchWaitQueenTask.start();
        testSource.catchWaitQueenTaskStatus=true;
    };
    cb(null,socketCons.toAll,'catchTicketsStatus',testSource.catchWaitQueenTaskStatus);
};

//查询取票任务状态
SocketControl.prototype.queryCatchTicketsStatus = function (bodyNode,cb) {
    cb(null,socketCons.toSelf,'queryCatchTicketsStatus',testSource.catchWaitQueenTaskStatus);
};


//查询用户列表
SocketControl.prototype.userList = function (bodyNode,cb) {
    db.collection('customer', {safe: true}, function (err, collection) {
        collection.find().toArray(function (err, datas) {
            cb(err,socketCons.toSelf,'userList',datas);
        })
    });
};

SocketControl.prototype.addUser = function (bodyNode,cb) {
    db.collection('customer', {safe: true}, function (err, collection) {
        collection.find({'userName': bodyNode.userName}).toArray(function (err, datas) {
           if(datas.length == 0){
               collection.insert(bodyNode, function () {
                   collection.find().toArray(function (err, datas) {
                       cb(err,socketCons.toAll,'userList',datas);
                   });
               });
           }else{
               collection.findAndModify({'userName': bodyNode.userName}, {'id': -1}, {$set: bodyNode}, {new: true}, function (err, result) {
                   collection.find().toArray(function (err, datas) {
                       cb(err,socketCons.toAll,'userList',datas);
                   });
               });
           };
        });
    });
};

SocketControl.prototype.terminalList = function (bodyNode,cb) {
    db.collection('terminal', {safe: true}, function (err, collection) {
        collection.find().toArray(function (err, datas) {
            cb(err,socketCons.toSelf,'terminalList',datas);
        })
    });
};


SocketControl.prototype.addTerminal = function (bodyNode,cb) {
    db.collection('terminal', {safe: true}, function (err, collection) {
        collection.insert(bodyNode, function () {
            db.createCollection('waitTerminal' + bodyNode.terminalName, {safe: true}, function () {
                collection.find().toArray(function (err, datas) {
                    console.log(datas);
                    cb(err,socketCons.toAll,'terminalList',datas);
                })
            });
        })
    })
};



module.exports = SocketControl;
