// print_control/TerminalControl.js
/**
 * Created by w44 on 15-1-8.
 */

var util = require('print_util');
var db = util.mongoDBUtil;

var cons = require('print_constants');
var terminalCons = cons.terminalCons;

var memory=require('print_memory');
var terminalArray = memory.terminalArray;

var bean=require('print_bean');
var ticketTask=bean.ticketTask;


var TerminalControl = function (socket, io) {
    var self = this;
    self.socket = socket;
    self.io = io;
};


TerminalControl.prototype.handle = function (headNode, bodyNode) {
    var self = this;
    console.log(JSON.stringify(headNode));
    console.log(JSON.stringify(bodyNode));
    var method = 'handle' + headNode.cmd;
    //添加判断方法逻辑，处理无此方法时的清空
    if (typeof self[method] === 'function') {
        return self[method](headNode, bodyNode);
    } else {
        console.log('找不到处理此命令的方法: ' + method);
        return null;
    }
};


/**
 * send response msg to terminal
 */
TerminalControl.prototype.sendRepMsg = function (backHeadNode) {
    var self = this;
    var content = backHeadNode.cmd + backHeadNode.sequenceId + backHeadNode.terminalId + backHeadNode.retCode + backHeadNode.retDesc;
    var contentBuf = new Buffer(content);
    //logger.info('----backHeadNode');
    //logger.info(backHeadNode);
    var buf = new Buffer(contentBuf.length + 4);
    buf.writeInt32BE(contentBuf.length, 0);
    contentBuf.copy(buf, 4, 0, contentBuf.length);
    self.sendBuf(buf);
};


/**
 * send buffer to terminal
 */
TerminalControl.prototype.sendBuf = function (buf) {
    var self = this;
    self.socket.write(buf);
};


/**
 *
 * @param headNode
 * @param bodyNode
 *
 * 2.1.7.1    终端机登记后请求数据消息(命令码:0001)
 * 请求消息: [终端机->服务器]
 *
 * 表示该机器处于可用状态,应发送出票请求
 * 得到该机器的ID号,修改状态为可用状态
 *
 */
TerminalControl.prototype.handle0001 = function (headNode, bodyNode) {
    var self = this;
    var backHeadNode = {
        cmd: headNode.cmd,
        sequenceId: headNode.sequenceId,
        retCode: terminalCons.retCode.success.code,
        retDesc: terminalCons.retCode.success.desc,
        terminalId: headNode.terminalId
    };
    var have = true;
    for (var i = 0; i < terminalArray.length; i++) {
        var terminalId = terminalArray[i].terminalId;
        if (terminalId == headNode.terminalId) {
            //todo 如果该终端机处于查询状态,也会发送0001,但是此时不应该做任何操作,但是需要更新数据库中终端机的状态
            if (terminalArray[i].status == terminalCons.terminal.status.search) {
                db.collection('terminal', {
                    safe: true
                }, function (err, collection) {
                    collection.findAndModify({
                        'id': headNode.terminalId
                    }, {}, {
                        $set: {
                            status: terminalCons.terminal.status.search,
                            address: self.socket.remoteAddress,
                            port: self.socket.remotePort
                        }
                    }, {
                        new: true
                    }, function (err, result) {
                        if (err) {
                            console.log(err);
                        } else {
                            //logger.info(result);
                            var msg = JSON.stringify(result);
                            msg.status=9000;
                            self.io.emit('statusChange', msg);
                            var _backHeadNode = {
                                cmd: '0001',
                                sequenceId: headNode.sequenceId,
                                retCode: terminalCons.retCode.success.code,
                                retDesc: terminalCons.retCode.success.desc,
                                terminalId: headNode.terminalId
                            };
                            self.sendRepMsg(_backHeadNode);
                        }
                    })
                });
                return;
            }
            terminalArray[i].status = terminalCons.terminal.status.available;
            terminalArray[i].terminal.status = terminalCons.terminal.status.available;
            //数组中有该终端
            have = false;
        }
    }
    //logger.info(flag);
    db.collection('terminal', {
        safe: true
    }, function (err, collection) {
        collection.findAndModify({
            'id': headNode.terminalId
        }, {}, {
            $set: {
                status: terminalCons.terminal.status.available,
                address: self.socket.remoteAddress,
                port: self.socket.remotePort
            }
        }, {
            new: true
        }, function (err, result) {
            if (err) {
                console.log('handle0001: '+err);
            } else {
                if (result) {
                    if (have) {
                        var terminal = new ticketTask(result, self);
                        terminalArray.push(terminal);
                        console.log('handle0001: '+'已指定终端机' + result.id + '为出票');
                    }
                    var msg = JSON.stringify(result);
                    self.io.emit('statusChange',msg);
                }else{
                    console.log('handle0001: 没有此编号的终端机：'+headNode.terminalId);
                };
            }
        })
    });
    //logger.info('数组长度 + :' + terminalArray.length);
    self.sendRepMsg(backHeadNode);
};


//发送出票请求后的响应消息
TerminalControl.prototype.handle0002 = function (headNode, bodyNode) {
    logger.info('----handle0002----');
    logger.info(headNode);
    logger.info(bodyNode);
};


//发送出票请求后收到的票面信息(如果成功)
TerminalControl.prototype.handle0003 = function (headNode, bodyNode) {
    var self = this;
    if (bodyNode == null) {
        logger.info('0003传来数据有误');
        return;
    } else if (bodyNode.data1 == '') {
        logger.info('0003传来数据有误');
        return;
    } else {
        logger.info('---' + headNode.cmd + '---');
        logger.info(headNode);
        logger.info(bodyNode);
        //todo 测试0005,暂时屏蔽0003
        //    if (headNode.cmd == '0003') {
        //        logger.info('暂时屏蔽0003');
        //        return;
        //    }
        var backHeadNode = {
            cmd: '0003',
            sequenceId: headNode.sequenceId,
            terminalId: headNode.terminalId
        };
        //入终端机成功库
        async.waterfall([
            function (cb) {
                //先获得终端机以及终端机对应的socket.
                var terminal;
                for (var i = 0; i < terminalArray.length; i++) {
                    if (headNode.terminalId == terminalArray[i].terminalId) {
                        terminal = terminalArray[i].terminal;
                    }
                }
                cb(null, terminal);
            },
            function (terminal, cb) {
                db.collection('waitTerminal' + terminal.terminalName, {
                    safe: true
                }, function (err, collection) {
                    //                for (var i = 0; i < bodyNode.data.length; i++) {
                    var ticketId = bodyNode.namedata1.split(/\W/g)[0];
                    var terminalId = headNode.terminalId;
                    //logger.info('---ticketid---');
                    //logger.info(ticketId);
                    if (bodyNode.data1 == '') {
                        logger.info(ticketId + '..........返回数据有误!');
                        return;
                    }
                    collection.findAndRemove({
                        'id': ticketId
                    }, [], function (err, result) {
                        if (!result) {
                            logger.info('票已被取走' + ticketId);
                        } else {

                            delete result['_id'];
                            result.terminalId = terminalId;
                            var arr = bodyNode.data1.split(/\n/);
                            if (result.gameCode == 'T51') {
                                //竞彩票面信息与其他彩种不同,需要专门处理
                                result.chuan = arr[3].replace(/\D/g, '');
                                var tar = arr[5].replace(/\s+/g, "|");
                                arr = tar.split('|');
                                result.ticketSeq = arr[0];
                                result.ticketPwd = arr[1];
                                result.ticketPwd2 = arr[2];
                            } else {
                                var tar = arr[1].replace(/\s+/g, "|");
                                arr = tar.split('|');
                                result.ticketSeq = arr[0];
                                result.ticketPwd = arr[1];
                                result.ticketPwd2 = arr[2];
                            }
                            result.terminalReturnTime = new Date().getTime();
                            result.metaTicket = bodyNode.data1;
                            result.metaTicketName = bodyNode.namedata1;
                            result.metaTicket2 = bodyNode.data2;
                            result.metaTicket2Name = bodyNode.namedata2;
                            result.status = prop.ticket.status.in;
                            result.platStatus = '';
                            //判断是否打印出的为样票
                            if (result.metaTicket.indexOf('样票') != -1) {
                                //打印出了一张样票,入错误库
                                db.collection('wrongTicket', {safe: true}, function (err, coll) {
                                    coll.insert(result, function () {
                                        //
                                        var msg = {
                                            id: result.id,
                                            msg: '打印出了一张样票!请至错误库查看这张票!'
                                        };
                                        self.io.emit('wrongTicket', msg);
                                    })
                                });
                                return;
                            }
                            try {
                                var flag = checkNumberUtil.checkMut(result, bodyNode.data1);
                                if (flag) {
                                    db.collection('TerminalPrintSuccess', {safe: true}, function (err, collection) {
                                        collection.insert(result, function (err, result) {
                                        })
                                    });
                                    //校验通过
                                    //todo 给平台发送出票成功请求
                                    platform.sendSuccess(result);
                                    if (headNode.cmd == '0051') {
                                        logger.info('---发送0053---');
                                        var backHeadNode0005 = {
                                            cmd: '0053',
                                            sequenceId: headNode.sequenceId,
                                            terminalid: terminalId,
                                            retCode: prop.retCode.success.code,
                                            retDesc: prop.retCode.success.desc
                                        };
                                        var backBodyNode0005 = {
                                            ticketid: result.id
                                        };
                                        self.send0053(backHeadNode0005, backBodyNode0005);
                                    } else if (headNode.cmd == '0031') {

                                        backHeadNode.retCode = prop.retCode.success.code;
                                        backHeadNode.retDesc = prop.retCode.success.desc;
                                        backHeadNode.ticketid = ticketId;

                                        self.sendRepMsg0003(backHeadNode);
                                    }
                                } else {
                                    //校验不通过

                                    platform.sendFail(result);
                                    db.collection('failTikcet', {safe: true}, function (err, colllection) {
                                        if (!err) {
                                            colllection.insert(result, function () {
                                                logger.info(result.id + '已入失败库');
                                            })
                                        }
                                    })
                                }


                            } catch (err) {
                                logger.info(err);
                                logger.info('校验过程中出错!');
                            }
                        }
                    });
                    //                }
                    //todo 此处做校验
                });
                cb(null);
            }
        ], function (err, result) {
            if (err) {
                logger.info('err: ', err); // -> null
                logger.info('result: ', result); // -> 16
            }
        })
    }

};


TerminalControl.prototype.handle0004 = function (headNode, bodyNode) {
    logger.info('---handle0004---');
    logger.info(bodyNode);
    var ticketId = bodyNode.ticketId;
    db.collection('TerminalPrintSuccess', {
        safe: true
    }, function (err, collection) {
        collection.findAndRemove({
            'id': ticketId
        }, [], function (err, result) {
            db.collection('History', {
                safe: true
            }, function (err, _collection) {

                delete result['_id'];
                result.printTime = new Date().getTime;
                result.bonusTime = new Date().getTime;

                _collection.insert(result, function (doc) {
                    logger.info(ticketId + '已进入历史库');
                })
            })
        })
    })
};


TerminalControl.prototype.handle0051 = function (headNode, bodyNode) {
    logger.info('---handle0051---');
    if (bodyNode == null) {
        logger.info('---handle0051BodyNode格式有误---');
        return;
    }
    var self = this;
    self.handle0003(headNode, bodyNode);
    //发送回执
    var body = {
        ticketid: bodyNode.namedata1.split(/\W/g)[0]
    };
    var backHeadNode = {
        cmd: '0053',
        sequenceId: headNode.sequenceId,
        terminalid: headNode.terminalId,
        retCode: prop.retCode.success.code,
        retDesc: prop.retCode.success.desc
    };
    self.send0053(backHeadNode, body);
};

TerminalControl.prototype.handle0052 = function (headNode, bodyNode) {
    logger.info('---handle0052---');
    logger.info(bodyNode);
};

TerminalControl.prototype.handle0061 = function (headNode, bodyNode) {
    logger.info('---handle0061---');
    logger.info(bodyNode.playId);
    //获得游戏代码
    var gameCode = gameCodeTrans.gameCodeTrans4Search[bodyNode.playId];
    var _date = moment().format('YYMMDD');
    //根据游戏名做具体处理
    _date = _date.substring(0, _date.length - 1);
    var bonusInfo = bodyNode.data.split('\r\n');
    bonusInfo.splice(0, 5);
    var maxTerm = {
        termCode: 0,
        number: ''
    };
    async.each(bonusInfo, function (item, cb) {
        if (gameCode == 'T05') {
            var inf = item.split(' ');
        } else if (gameCode == 'T06') {
            var inf = item.replace(/\s+/g, ' ').split(' ');
        }

        if (inf[1]) {
            var firstNum = {};
            firstNum.termCode = parseInt(_date + inf[0]);
            firstNum.number = inf[1].replace(/\D/g, ',');
            if (firstNum.termCode > maxTerm.termCode) {
                maxTerm = firstNum;
            }
        }
        if (inf[3]) {
            var secondNum = {};
            secondNum.termCode = parseInt(_date + inf[2]);
            secondNum.number = inf[3].replace(/\D/g, ',');
            if (secondNum.termCode > maxTerm.termCode) {
                maxTerm = secondNum;
            }
        }
        cb(null);
    }, function (err) {
        if (maxTerm.number == '') {
            numberCtrl.query(gameCode)
        } else {
            numberCtrl.sendNumbers2plat(gameCode, maxTerm.termCode, maxTerm.number);
            db.collection('WinNumbers', {safe: true}, function (err, collection) {
                maxTerm.gameCode = gameCode;
                maxTerm.writeTime = new Date().getTime();
                if (!err) {
                    collection.insert(maxTerm, function () {
                    })
                }
            });
            numberCtrl.run(gameCode);
        }
    });
};


TerminalControl.prototype.handle1000 = function (headNode, bodyNode) {
    var self = this;
    logger.info('---handle1000---');
    logger.info(headNode);
    logger.info(bodyNode);
    var terminalId = headNode.terminalId;
    async.waterfall([
        function (cb) {
            var have = true;
            for (var i = 0; i < terminalArray.length; i++) {
                if (terminalId == terminalArray[i].terminalId) {
                    terminalArray[i].status = prop.terminal.status.bonus;
                    terminalArray[i].terminal.status = prop.terminal.status.bonus;
                    have = false;
                }
            }
            cb(null, terminalId, have);
        },
        function (terminalId, have, cb) {
            db.collection('terminal', {
                safe: true
            }, function (err, collection) {
                collection.findAndModify({
                    'id': terminalId
                }, {
                    'id': -1
                }, {
                    $set: {
                        status: prop.terminal.status.bonus
                    }
                }, {
                    new: true
                }, function (err, result) {
                    var msg = JSON.stringify(result);
                    self.io.emit('statusChange', msg);
                    //logger.info('已指定终端机' + result.id + '为兑奖');
                    if (have) {
                        var terminal = new ticketTask(result, self);
                        terminalArray.push(terminal);
                    }
                    cb(null);
                })
            })
        }
    ], function (err, result) {
        if (err) {
            logger.info('err: ', err); // -> null
            logger.info('result: ', result); // -> 16
        }
    });


    var backHeadNode = {
        cmd: headNode.cmd,
        sequenceId: headNode.sequenceId,
        retCode: prop.retCode.success.code,
        retDesc: prop.retCode.success.desc,
        terminalId: headNode.terminalId
    };
    self.sendRepMsg(backHeadNode);
};


TerminalControl.prototype.handle2000 = function (headNode, bodyNode) {
    var self = this;
    //logger.info('---handle2000---');
    //logger.info(headNode);
    //logger.info(bodyNode);
    var terminalId = headNode.terminalId;
    async.waterfall([
        function (cb) {
            var have = true;
            for (var i = 0; i < terminalArray.length; i++) {
                if (terminalId == terminalArray[i].terminalId) {
                    terminalArray[i].status = prop.terminal.status.print;
                    terminalArray[i].terminal.status = prop.terminal.status.print;
                    have = false;
                }
            }
            cb(null, terminalId, have);
        },
        function (terminalId, have, cb) {
            db.collection('terminal', {
                safe: true
            }, function (err, collection) {
                collection.findAndModify({
                    'id': terminalId
                }, {
                    'id': -1
                }, {
                    $set: {
                        status: prop.terminal.status.print
                    }
                }, {
                    new: true
                }, function (err, result) {
                    var msg = JSON.stringify(result);
                    self.io.emit('statusChange', msg);
                    //logger.info('已指定终端机' + result.id + '为打印');
                    if (have) {
                        var terminal = new ticketTask(result, self);
                        terminalArray.push(terminal);
                    }
                    cb(null);
                })
            })
        }
    ], function (err, result) {
        if (err) {
            logger.info('err: ', err); // -> null
            logger.info('result: ', result); // -> 16
        }
    });


    var backHeadNode = {
        cmd: headNode.cmd,
        sequenceId: headNode.sequenceId,
        retCode: prop.retCode.success.code,
        retDesc: prop.retCode.success.desc,
        terminalId: headNode.terminalId
    };
    self.sendRepMsg(backHeadNode);
};

TerminalControl.prototype.handle0071 = function (headNode, bodyNode) {
    var self = this;
    logger.info('---handle0071---');
    logger.info(bodyNode.data);
    var arr = bodyNode.data.split('$');
    logger.info('数组长度' + arr.length);
    db.collection('TerminalPrintSuccess', {
        safe: true
    }, function (err, collection) {
        for (var i = 0; i < arr.length; i++) {
            var _bonusInfo = arr[i];
            logger.info('_bonusInfo + .........' + _bonusInfo);
            if (_bonusInfo.indexOf('奖期不在可兑奖状态') >= 0) { //不在兑奖时间内
                var seq = _bonusInfo.substring(0, 20);
                collection.findAndModify({
                    'ticketSeq': seq
                }, {
                    'id': -1
                }, {
                    $set: {
                        status: prop.ticket.status.in
                    }
                }, {
                    new: true
                }, function (err, result) {
                    logger.info(seq + "返回兑奖状态为未兑奖,修改状态为等待兑奖")
                });
                continue;
            }
            if (_bonusInfo.length > 70) {
                var seq = _bonusInfo.substring(0, 20);
                logger.info('seq + .........' + seq);
                self.insert2History(headNode.terminalId, _bonusInfo, seq);
            } else {
                var seq = _bonusInfo.substring(0, 20);
                if (seq.length == 20) {
                    collection.findAndModify({
                        'ticketSeq': seq
                    }, {
                        'id': -1
                    }, {
                        $set: {
                            status: prop.ticket.status.in
                        }
                    }, {
                        new: true
                    }, function (err, result) {
                        logger.info(seq + "未返回兑奖状态,修改状态为等待兑奖")
                    })
                }
            }
        }
    })
};


TerminalControl.prototype.handle0072 = function (headNode, bodyNode) {
    logger.info('---handle0072---');
    logger.info(bodyNode.data);
};


TerminalControl.prototype.handle0032 = function (headNode, bodyNode) {
    var self = this;
    logger.info('---handle' + headNode.cmd + '---');
    //0032/0033用于处理错误票,此时应先将该票放入失败库,然后通知平台票出失败了
    //首先获取该票所在的库
    logger.info(headNode);
    logger.info(bodyNode);
    if (!bodyNode.ticketId) {
        return;
    }
    var ticketId = bodyNode.ticketId;

    db.collection('terminal', {safe: true}, function (err, collection) {
        collection.findOne({'id': headNode.terminalId}, function (err, terminal) {
            if (terminal) {
                var name = 'waitTerminal' + terminal.terminalName;
                db.collection(name, {safe: true}, function (err, _coll) {
                    _coll.findAndRemove({'id': ticketId}, [], function (err, _ticket) {
                        db.collection('failTicket', {safe: true}, function (err, __coll) {
                            if (!_ticket) {
                                logger.info('不存在的票据 : ' + bodyNode.ticketId);
                            } else {
                                //通知平台失败
                                platform.sendFail(_ticket);
                                //入失败库
                                delete _ticket['_id'];
                                //新增失败信息,失败代码
                                _ticket.failInfo = bodyNode.retCode;
                                _ticket.failType = bodyNode.retDesc;
                                _ticket.terminalReturnTime = new Date().getTime();

                                __coll.insert(_ticket, function () {
                                });
                                if (headNode.cmd == '0033') {
                                    logger.info('---发送0055---');
                                    var backHeadNode0005 = {
                                        cmd: '0055',
                                        sequenceId: headNode.sequenceId,
                                        terminalid: headNode.terminalId,
                                        retCode: prop.retCode.success.code,
                                        retDesc: prop.retCode.success.desc
                                    };
                                    var backBodyNode0005 = {
                                        ticketid: _ticket.id
                                    };
                                    self.send0053(backHeadNode0005, backBodyNode0005);
                                }
                            }
                        })
                    })
                })
            }
        });
    })
};


TerminalControl.prototype.handle0054 = function (headNode, bodyNode) {
    var self = this;
    logger.info('---handle' + headNode.cmd + '---');
    //0032/0033用于处理错误票,此时应先将该票放入失败库,然后通知平台票出失败了
    //首先获取该票所在的库
    logger.info(bodyNode.retCode);
    logger.info(bodyNode.retDesc);
    self.failHandler(headNode, bodyNode);
};


TerminalControl.prototype.handle4000 = function (headNode, bodyNode) {
    var self = this;
    logger.info('----handle4000----');
    logger.info(headNode);
    //终端机要死机,先从等待返回库中移走这些票,再改数组中状态,最后修改数据库的状态
    //需要消息回执
    var terminalId = headNode.terminalId;
    //step 1,从等待返回库中移走这些票
    for (var i = 0; i < terminalArray.length; i++) {
        var terminal = terminalArray[i];
        if (terminal.terminal.id == terminalId) {
            terminalArray.splice(i, 1);
        }
    }
    self.deadHandler(headNode);
};


TerminalControl.prototype.deadHandler = function (headNode) {
    var self = this;
    //打印模式死机
    db.collection('terminal', {safe: true}, function (err, collection) {
        collection.findAndModify({id: headNode.terminalId}, {}, {$set: {status: prop.terminal.status.notAvailable}}, {new: true}, function (err, terminal) {
            //通知界面
            if (terminal) {
                var msg = JSON.stringify(terminal);
                self.io.emit('statusChange', msg);
                db.collection('waitTerminal' + terminal.terminalName, {safe: true}, function (err, _coll) {
                    _coll.find({}).toArray(function (err, tickets) {
                        async.each(tickets, function (ticket, callback) {
                            db.collection('waitTerminalTicket', {safe: true}, function (err, _col) {
                                delete ticket['_id'];
                                _col.insert(ticket, function () {
                                    //已添加到等待取走库
                                    callback(null);
                                })
                            })
                        }, function (err) {
                            var _backHeadNode = {
                                cmd: '4000',
                                sequenceId: headNode.sequenceId,
                                retCode: prop.retCode.success.code,
                                retDesc: prop.retCode.success.desc,
                                terminalId: headNode.terminalId
                            };
                            self.sendRepMsg(_backHeadNode);
                        })
                    });
                    _coll.remove({}, {safe: true}, function () {
                    })
                })
            } else {
                logger.info('找不到的终端机' + headNode.terminalId);
            }
        })
    })
};

TerminalControl.prototype.connClose = function (remoteAddress) {
    var self = this;
    //把状态修改为不可用,从数组中移除.
    for (var i = 0; i < terminalArray.length; i++) {
        var terminal = terminalArray[i];
        if (terminal.terminal.address == remoteAddress) {
            terminalArray[i].status = terminalCons.terminal.status.notAvailable;
            terminalArray.splice(i, 1);
            db.collection('terminal', {
                safe: true
            }, function (err, collection) {
                collection.findAndModify({
                    'id': terminal.terminal.id
                }, {
                    'id': -1
                }, {
                    $set: {
                        status: terminalCons.terminal.status.notAvailable
                    }
                }, {
                    new: true
                }, function (err, result) {
                    var msg = JSON.stringify(result);
                    self.io.emit('statusChange', msg);
                    console.log('更改终端机' + result.id + '状态为' + result.status);
                })
            })
        }
    }
};


TerminalControl.prototype.insert2History = function (terminalId, bonusInfo, seq) {
    db.collection('TerminalPrintSuccess', {
        safe: true
    }, function (err, collection) {
        collection.findAndRemove({
            'ticketSeq': seq
        }, [], function (err, result) {
            if (result) {
                var bonus = 0;
                delete result['_id'];
                result.bonusCount = 0;
                if (bonusInfo.indexOf('请到体育彩票管理中心领奖') != -1) {
                    var arr = bonusInfo.split('\r\n');
                    async.each(arr, function (item) {
                        if (item.indexOf('合计') != -1) {
                            bonus = parseInt(item.replace(/\D/g, ''));
                        }
                    }, function (err) {
                    });


                    result.bonusInfo = bonusInfo;
                    result.status = prop.ticket.status.bigBonus;
                    result.bonus = bonus;
                    result.bonusCount = 1;
                    collection.insert(result, function () {
                        logger.info(result.id + '已中大奖');
                    })
                } else {
                    if (bonusInfo.indexOf('恭喜中奖') != -1) {
                        result.bonusCount = 1;
                        var arr = bonusInfo.split('\r\n');
                        async.each(arr, function (item) {
                            if (item.indexOf('合计') != -1) {
                                bonus = parseInt(item.replace(/\D/g, ''));
                            }
                        }, function (err) {
                        })
                    }
                    result.bonusTime = new Date().getTime();
                    result.bonus = bonus;
                    result.bonusInfo = bonusInfo;
                    result.bonusTerminalId = terminalId;
                    db.collection('History', {
                        safe: true
                    }, function (err, _coll) {
                        _coll.insert(result, function () {
                            logger.info(result.ticketSeq + '已入历史库');
                        })
                    })
                }
            } else {
                logger.info(err);
            }
        })
    });
};

TerminalControl.prototype.failHandler = function (headNode, bodyNode) {
    var self = this;
    var ticketId = bodyNode.ticketId;
    db.collection('terminal', {safe: true}, function (err, collection) {
        collection.findOne({'id': headNode.terminalId}, function (err, terminal) {
            if (terminal) {
                db.collection('waitTerminal' + terminal.terminalName, {safe: true}, function (err, _coll) {
                    _coll.findAndRemove({'id': ticketId}, [], function (err, ticket) {
                        if (ticket) {
                            db.collection('failTicket', {safe: true}, function (err, __coll) {
                                logger.info(ticketId);
                                //通知平台失败
                                var body = {};
                                platform.sendFail(ticket);
                                delete ticket['_id'];
                                //新增失败信息,失败代码
                                ticket.failInfo = bodyNode.retCode;
                                ticket.failType = bodyNode.retDesc;
                                ticket.terminalReturnTime = new Date().getTime();
                                //入失败库
                                __coll.insert(ticket, function () {
                                });
                                //通知终端机失败
                                logger.info('---发送0055---');
                                var backHeadNode0005 = {
                                    cmd: '0055',
                                    sequenceId: headNode.sequenceId,
                                    terminalid: headNode.terminalId,
                                    retCode: prop.retCode.success.code,
                                    retDesc: prop.retCode.success.desc
                                };
                                var backBodyNode0005 = {
                                    ticketid: ticket.id
                                };
                                self.send0053(backHeadNode0005, backBodyNode0005);
                            })
                        }
                    })
                })
            }
        });
    })
};
TerminalControl.prototype.sendSuccess2Plat = function (message) {

    platform.get(prop.platform.site.key, prop.platform.command.ticketResponse, message, prop.platform.site.channelCode, function (err, body) {
        if (err) {
            TerminalControl.prototype.sendSuccess2Plat(message);
        } else {
            var _body = JSON.parse(body.body);
            db.collection('TerminalPrintSuccess', {safe: true}, function (err, collection) {
                collection.update({id: message.ticketId}, {
                    $set: {platStatus: _body.repCode}
                }, {}, function (err, result) {
                    logger.info(message.ticketId + '平台状态 : ' + _body.repCode);
                })
            });
        }
    });
};

TerminalControl.prototype.sendFail2Plat = function (message) {
    platform.get(prop.platform.site.key, prop.platform.command.ticketResponse, message, prop.platform.site.channelCode, function (err, body) {
        if (err) {
            TerminalControl.prototype.sendFail2Plat(message);
        } else {
            logger.info(body);
        }
    });
};


module.exports = TerminalControl;
