// print_control/TerminalControl.js
/**
 * Created by w44 on 15-1-8.
 */

var async = require('async');

var util = require('print_util');
var mongoDBUtil = util.mongoDBUtil;
var log = util.log;
var checkNumberUtil = util.checkNumberUtil;

var assist = require('print_assist');
var nodePlatAssist = assist.nodePlatAssist;

var cons = require('print_constants');
var terminalCons = cons.terminalCons;
var msgParam = cons.msgParam;
var nodePlatCons = cons.nodePlatCons;

var memory = require('print_memory');
var terminalArray = memory.terminalArray;

var bean = require('print_bean');
var ticketTask = bean.ticketTask;


var TerminalControl = function (socket, io) {
    var self = this;
    self.socket = socket;
    self.io = io;
};


TerminalControl.prototype.handle = function (headNode, bodyNode) {
    var self = this;
    console.log(JSON.stringify(headNode));
    console.log(JSON.stringify(bodyNode));
    var method = 'handle' + headNode.cmd;
    //特殊处理
    if (headNode.cmd == '0031' || headNode.cmd == '0051') {
        method = 'handle' + '0003';
    }
    //添加判断方法逻辑，处理无此方法时的清空
    if (typeof self[method] === 'function') {
        self[method](headNode, bodyNode);
    } else {
        log.info('找不到处理此命令的方法: ' + method);
    }
};


/**
 * send response msg to terminal
 */
TerminalControl.prototype.sendRepMsg = function (backHeadNode) {
    var self = this;
    var content = backHeadNode.cmd + backHeadNode.sequenceId + backHeadNode.terminalId + backHeadNode.retCode + backHeadNode.retDesc;
    var contentBuf = new Buffer(content);
    var buf = new Buffer(contentBuf.length + 4);
    buf.writeInt32BE(contentBuf.length, 0);
    contentBuf.copy(buf, 4, 0, contentBuf.length);
    self.sendBuf(buf);
};

/**
 * send response 0003 to terminal
 */

TerminalControl.prototype.sendRepMsg0003 = function (backHeadNode) {
    var self = this;
    var content = backHeadNode.cmd + backHeadNode.sequenceId + backHeadNode.terminalId + backHeadNode.retCode + backHeadNode.retDesc + backHeadNode.ticketid;
    var contentBuf = new Buffer(content);
    var buf = new Buffer(contentBuf.length + 4);
    buf.writeInt32BE(contentBuf.length, 0);
    contentBuf.copy(buf, 4, 0, contentBuf.length);
    self.sendBuf(buf);
    log.info('---发送0003---');
    log.info(backHeadNode);
};


/**
 * send buffer to terminal
 */
TerminalControl.prototype.sendBuf = function (buf) {
    var self = this;
    self.socket.write(buf);
};


/**
 *
 * @param headNode
 * @param bodyNode
 *
 * 2.1.7.1    终端机登记后请求数据消息(命令码:0001)
 * 请求消息: [终端机->服务器]
 *
 * 表示该机器处于可用状态,应发送出票请求
 * 得到该机器的ID号,修改状态为可用状态
 *
 */
TerminalControl.prototype.handle0001 = function (headNode, bodyNode) {
    var self = this;
    var backHeadNode = {
        cmd: headNode.cmd,
        sequenceId: headNode.sequenceId,
        retCode: terminalCons.retCode.success.code,
        retDesc: terminalCons.retCode.success.desc,
        terminalId: headNode.terminalId
    };
    var have = true;
    for (var i = 0; i < terminalArray.length; i++) {
        var terminalId = terminalArray[i].terminalId;
        if (terminalId == headNode.terminalId) {
            //todo 如果该终端机处于查询状态,也会发送0001,但是此时不应该做任何操作,但是需要更新数据库中终端机的状态
            if (terminalArray[i].status == terminalCons.terminal.status.search) {
                mongoDBUtil.db.collection('terminal', {
                    safe: true
                }, function (err, collection) {
                    collection.findAndModify({
                        'id': headNode.terminalId
                    }, {}, {
                        $set: {
                            status: terminalCons.terminal.status.search,
                            address: self.socket.remoteAddress,
                            port: self.socket.remotePort
                        }
                    }, {
                        new: true
                    }, function (err, result) {
                        if (err) {
                            console.log(err);
                        } else {
                            log.info(result);
                            var msg = JSON.stringify(result);
                            self.io.emit('statusChange', msg);
                            var _backHeadNode = {
                                cmd: '0001',
                                sequenceId: headNode.sequenceId,
                                retCode: terminalCons.retCode.success.code,
                                retDesc: terminalCons.retCode.success.desc,
                                terminalId: headNode.terminalId
                            };
                            self.sendRepMsg(_backHeadNode);
                        }
                    })
                });
                return;
            }
            terminalArray[i].status = terminalCons.terminal.status.available;
            terminalArray[i].terminal.status = terminalCons.terminal.status.available;
            //数组中有该终端
            have = false;
        }
    }
    mongoDBUtil.db.collection('terminal', {
        safe: true
    }, function (err, collection) {
        collection.findAndModify({
            'id': headNode.terminalId
        }, {}, {
            $set: {
                status: terminalCons.terminal.status.available,
                address: self.socket.remoteAddress,
                port: self.socket.remotePort
            }
        }, {
            new: true
        }, function (err, result) {
            if (err) {
                log.error('handle0001: ' + err);
            } else {
                if (result) {
                    if (have) {
                        var terminal = new ticketTask(result, self);
                        terminalArray.push(terminal);
                        log.info('handle0001: ' + '已指定终端机' + result.id + '为出票');
                    }
                    var msg = JSON.stringify(result);
                    //通知页面修改
                    self.io.emit('statusChange', msg);
                } else {
                    log.error('handle0001: 没有此编号的终端机：' + headNode.terminalId);
                }
                ;
            }
        })
    });
    self.sendRepMsg(backHeadNode);
};

//发送出票请求后的响应消息
TerminalControl.prototype.handle0002 = function (headNode, bodyNode) {
    log.info('----handle0002----');
    log.info(headNode);
    log.info(bodyNode);
};


//发送出票请求后收到的票面信息(如果成功)
TerminalControl.prototype.handle0003 = function (headNode, bodyNode) {
    var self = this;
    if (bodyNode == null) {
        log.error('0003传来数据有误');
        return;
    } else if (bodyNode.data1 == '') {
        log.error('0003传来数据有误');
        return;
    } else {
        log.info('---' + headNode.cmd + '---');
        var backHeadNode = {
            cmd: '0003',
            sequenceId: headNode.sequenceId,
            terminalId: headNode.terminalId
        };
        //入终端机成功库
        async.waterfall([
            function (cb) {
                //先获得终端机以及终端机对应的socket.
                var terminal;
                for (var i = 0; i < terminalArray.length; i++) {
                    if (headNode.terminalId == terminalArray[i].terminalId) {
                        terminal = terminalArray[i].terminal;
                    }
                }
                cb(null, terminal);
            },
            function (terminal, cb) {
                mongoDBUtil.db.collection('waitTerminal' + terminal.terminalName, {
                    safe: true
                }, function (err, collection) {
                    var ticketId = bodyNode.namedata1.split(/\W/g)[0];
                    var terminalId = headNode.terminalId;
                    if (bodyNode.data1 == '') {
                        log.info(ticketId + '..........返回数据有误!');
                        return;
                    }
                    log.error('出票成功，从等待中取票并删除：' + ticketId);
                    collection.findAndRemove({
                        'id': ticketId
                    }, [], function (err, result) {
                        if (!result) {
                            log.info('票已被取走' + ticketId);
                        } else {
                            delete result['_id'];
                            result.terminalId = terminalId;
                            var arr = bodyNode.data1.split(/\n/);
                            if (result.gameCode == 'T51') {
                                //竞彩票面信息与其他彩种不同,需要专门处理
                                result.chuan = arr[3].replace(/\D/g, '');
                                var tar = arr[5].replace(/\s+/g, "|");
                                arr = tar.split('|');
                                result.ticketSeq = arr[0];
                                result.ticketPwd = arr[1];
                                result.ticketPwd2 = arr[2];
                            } else {
                                var tar = arr[1].replace(/\s+/g, "|");
                                arr = tar.split('|');
                                result.ticketSeq = arr[0];
                                result.ticketPwd = arr[1];
                                result.ticketPwd2 = arr[2];
                            }
                            result.terminalReturnTime = new Date().getTime();
                            result.metaTicket = bodyNode.data1;
                            result.metaTicketName = bodyNode.namedata1;
                            result.metaTicket2 = bodyNode.data2;
                            result.metaTicket2Name = bodyNode.namedata2;
                            result.status = terminalCons.ticket.status.in;
                            result.platStatus = '';
                            //判断是否打印出的为样票
                            if (result.metaTicket.indexOf('样票') != -1) {
                                //打印出了一张样票,入错误库
                                mongoDBUtil.db.collection('wrongTicket', {safe: true}, function (err, coll) {
                                    coll.insert(result, function () {
                                        var msg = {
                                            id: result.id,
                                            msg: '打印出了一张样票!请至错误库查看这张票!'
                                        };
                                        self.io.emit('wrongTicket', msg);
                                    })
                                });
                                return;
                            };
                            try {
                                var flag = checkNumberUtil.checkMut(result, bodyNode.data1);
                                if (flag) {
                                    collection.count({}, {}, function (err, count) {
                                        log.error(count);
                                        if(count<1){

                                            for (var i = 0; i < terminalArray.length; i++) {
                                                if (headNode.terminalId == terminalArray[i].terminalId) {
                                                    terminalArray[i].status = terminalCons.terminal.status.available;
                                                    terminalArray[i].terminal.status = terminalCons.terminal.status.available;
                                                    var msg = JSON.stringify(terminalArray[i].terminal);
                                                    self.io.emit('statusChange', msg);
                                                }
                                            }
                                        }
                                    });

                                    mongoDBUtil.db.collection('TerminalPrintSuccess', {safe: true}, function (err, collection) {
                                        collection.insert(result, function (err, result) {
                                        })
                                    });
                                    //校验通过
                                    //todo 给平台发送出票成功请求
                                    nodePlatAssist.sentP02(result, function (err, backMsgNode) {
                                        //todo  处理平台返回
                                        if (!err) {
                                            var _body = JSON.parse(backMsgNode.body);
                                            mongoDBUtil.db.collection('TerminalPrintSuccess', {safe: true}, function (err, collection) {
                                                collection.update({id: result.id}, {
                                                    $set: {platStatus: _body.repCode}
                                                }, {}, function (err, res) {
                                                    log.info(result.id + '平台状态 : ' + _body.repCode);
                                                })
                                            });
                                        }
                                    });
                                    if (headNode.cmd == '0051') {
                                        log.info('---发送0053---');
                                        var backHeadNode0005 = {
                                            cmd: '0053',
                                            sequenceId: headNode.sequenceId,
                                            terminalid: terminalId,
                                            retCode: nodePlatCons.retCode.success.code,
                                            retDesc: nodePlatCons.retCode.success.desc
                                        };
                                        var backBodyNode0005 = {
                                            ticketid: result.id
                                        };
                                        self.send0053(backHeadNode0005, backBodyNode0005);
                                    } else if (headNode.cmd == '0031') {
                                        backHeadNode.retCode = nodePlatCons.retCode.success.code;
                                        backHeadNode.retDesc = nodePlatCons.retCode.success.desc;
                                        backHeadNode.ticketid = ticketId;
                                        self.sendRepMsg0003(backHeadNode);
                                    }
                                } else {
                                    //校验不通过
                                    platform.sendFail(result);
                                    mongoDBUtil.db.collection('failTikcet', {safe: true}, function (err, colllection) {
                                        if (!err) {
                                            colllection.insert(result, function () {
                                                log.error(result.id + '已入失败库');
                                            })
                                        }
                                    })
                                }
                            } catch (err) {
                                log.error('校验过程中出错!');
                                log.error(err);
                            }
                        }
                    });
                });
                cb(null);
            }
        ], function (err, result) {
            if (err) {
                log.error('err: ', err); // -> null
            }
        })
    }

};


TerminalControl.prototype.handle0004 = function (headNode, bodyNode) {
    log.info('---handle0004---');
    var ticketId = bodyNode.ticketId;
    mongoDBUtil.db.collection('TerminalPrintSuccess', {
        safe: true
    }, function (err, collection) {
        collection.findAndRemove({
            'id': ticketId
        }, [], function (err, result) {
            mongoDBUtil.db.collection('History', {
                safe: true
            }, function (err, _collection) {

                delete result['_id'];
                result.printTime = new Date().getTime;
                result.bonusTime = new Date().getTime;

                _collection.insert(result, function (doc) {
                    log.info(ticketId + '已进入历史库');
                })
            })
        })
    })
};


TerminalControl.prototype.handle0051 = function (headNode, bodyNode) {
    log.info('---handle0051---');
    if (bodyNode == null) {
        log.error('---handle0051BodyNode格式有误---');
        return;
    }
    var self = this;
    self.handle0003(headNode, bodyNode);
    //发送回执
    var body = {
        ticketid: bodyNode.namedata1.split(/\W/g)[0]
    };
    var backHeadNode = {
        cmd: '0053',
        sequenceId: headNode.sequenceId,
        terminalid: headNode.terminalId,
        retCode: prop.retCode.success.code,
        retDesc: prop.retCode.success.desc
    };
    self.send0053(backHeadNode, body);
};

TerminalControl.prototype.handle0052 = function (headNode, bodyNode) {
    logger.info('---handle0052---');
    logger.info(bodyNode);
};

TerminalControl.prototype.handle0061 = function (headNode, bodyNode) {
    logger.info('---handle0061---');
    logger.info(bodyNode.playId);
    //获得游戏代码
    var gameCode = gameCodeTrans.gameCodeTrans4Search[bodyNode.playId];
    var _date = moment().format('YYMMDD');
    //根据游戏名做具体处理
    _date = _date.substring(0, _date.length - 1);
    var bonusInfo = bodyNode.data.split('\r\n');
    bonusInfo.splice(0, 5);
    var maxTerm = {
        termCode: 0,
        number: ''
    };
    async.each(bonusInfo, function (item, cb) {
        if (gameCode == 'T05') {
            var inf = item.split(' ');
        } else if (gameCode == 'T06') {
            var inf = item.replace(/\s+/g, ' ').split(' ');
        }

        if (inf[1]) {
            var firstNum = {};
            firstNum.termCode = parseInt(_date + inf[0]);
            firstNum.number = inf[1].replace(/\D/g, ',');
            if (firstNum.termCode > maxTerm.termCode) {
                maxTerm = firstNum;
            }
        }
        if (inf[3]) {
            var secondNum = {};
            secondNum.termCode = parseInt(_date + inf[2]);
            secondNum.number = inf[3].replace(/\D/g, ',');
            if (secondNum.termCode > maxTerm.termCode) {
                maxTerm = secondNum;
            }
        }
        cb(null);
    }, function (err) {
        if (maxTerm.number == '') {
            numberCtrl.query(gameCode)
        } else {
            numberCtrl.sendNumbers2plat(gameCode, maxTerm.termCode, maxTerm.number);
            mongoDBUtil.db.collection('WinNumbers', {safe: true}, function (err, collection) {
                maxTerm.gameCode = gameCode;
                maxTerm.writeTime = new Date().getTime();
                if (!err) {
                    collection.insert(maxTerm, function () {
                    })
                }
            });
            numberCtrl.run(gameCode);
        }
    });
};


TerminalControl.prototype.handle1000 = function (headNode, bodyNode) {
    var self = this;
    logger.info('---handle1000---');
    logger.info(headNode);
    logger.info(bodyNode);
    var terminalId = headNode.terminalId;
    async.waterfall([
        function (cb) {
            var have = true;
            for (var i = 0; i < terminalArray.length; i++) {
                if (terminalId == terminalArray[i].terminalId) {
                    terminalArray[i].status = prop.terminal.status.bonus;
                    terminalArray[i].terminal.status = prop.terminal.status.bonus;
                    have = false;
                }
            }
            cb(null, terminalId, have);
        },
        function (terminalId, have, cb) {
            mongoDBUtil.db.collection('terminal', {
                safe: true
            }, function (err, collection) {
                collection.findAndModify({
                    'id': terminalId
                }, {
                    'id': -1
                }, {
                    $set: {
                        status: prop.terminal.status.bonus
                    }
                }, {
                    new: true
                }, function (err, result) {
                    var msg = JSON.stringify(result);
                    self.io.emit('statusChange', msg);
                    //logger.info('已指定终端机' + result.id + '为兑奖');
                    if (have) {
                        var terminal = new ticketTask(result, self);
                        terminalArray.push(terminal);
                    }
                    cb(null);
                })
            })
        }
    ], function (err, result) {
        if (err) {
            logger.info('err: ', err); // -> null
            logger.info('result: ', result); // -> 16
        }
    });


    var backHeadNode = {
        cmd: headNode.cmd,
        sequenceId: headNode.sequenceId,
        retCode: prop.retCode.success.code,
        retDesc: prop.retCode.success.desc,
        terminalId: headNode.terminalId
    };
    self.sendRepMsg(backHeadNode);
};


TerminalControl.prototype.handle0071 = function (headNode, bodyNode) {
    var self = this;
    logger.info('---handle0071---');
    logger.info(bodyNode.data);
    var arr = bodyNode.data.split('$');
    logger.info('数组长度' + arr.length);
    mongoDBUtil.db.collection('TerminalPrintSuccess', {
        safe: true
    }, function (err, collection) {
        for (var i = 0; i < arr.length; i++) {
            var _bonusInfo = arr[i];
            logger.info('_bonusInfo + .........' + _bonusInfo);
            if (_bonusInfo.indexOf('奖期不在可兑奖状态') >= 0) { //不在兑奖时间内
                var seq = _bonusInfo.substring(0, 20);
                collection.findAndModify({
                    'ticketSeq': seq
                }, {
                    'id': -1
                }, {
                    $set: {
                        status: prop.ticket.status.in
                    }
                }, {
                    new: true
                }, function (err, result) {
                    logger.info(seq + "返回兑奖状态为未兑奖,修改状态为等待兑奖")
                });
                continue;
            }
            if (_bonusInfo.length > 70) {
                var seq = _bonusInfo.substring(0, 20);
                logger.info('seq + .........' + seq);
                self.insert2History(headNode.terminalId, _bonusInfo, seq);
            } else {
                var seq = _bonusInfo.substring(0, 20);
                if (seq.length == 20) {
                    collection.findAndModify({
                        'ticketSeq': seq
                    }, {
                        'id': -1
                    }, {
                        $set: {
                            status: prop.ticket.status.in
                        }
                    }, {
                        new: true
                    }, function (err, result) {
                        logger.info(seq + "未返回兑奖状态,修改状态为等待兑奖")
                    })
                }
            }
        }
    })
};


/**
 * 处理打印模式终端机心跳数据
 *
 * @param headNode
 * @param bodyNode
 */

TerminalControl.prototype.handle2000 = function (headNode, bodyNode) {
    var self = this;
    var terminalId = headNode.terminalId;
    async.waterfall([
        function (cb) {
            var have = true;
            for (var i = 0; i < terminalArray.length; i++) {
                if (terminalId == terminalArray[i].terminalId) {
                    terminalArray[i].status = terminalCons.terminal.status.print;
                    terminalArray[i].terminal.status = terminalCons.terminal.status.print;
                    have = false;
                }
            }
            cb(null, terminalId, have);
        },
        function (terminalId, have, cb) {
            mongoDBUtil.db.collection('terminal', {
                safe: true
            }, function (err, collection) {
                collection.findAndModify({
                    'id': terminalId
                }, {
                    'id': -1
                }, {
                    $set: {
                        status: terminalCons.terminal.status.print
                    }
                }, {
                    new: true
                }, function (err, result) {
                    var msg = JSON.stringify(result);
                    self.io.emit('statusChange', msg);
                    //log.info('已指定终端机' + result.id + '为打印');
                    if (have) {
                        var terminal = new ticketTask(result, self);
                        terminalArray.push(terminal);
                    }
                    cb(null);
                })
            })
        }
    ], function (err, result) {
        if (err) {
            log.error('err: ', err); // -> null
        }
    });
    var backHeadNode = {
        cmd: headNode.cmd,
        sequenceId: headNode.sequenceId,
        retCode: terminalCons.retCode.success.code,
        retDesc: terminalCons.retCode.success.desc,
        terminalId: headNode.terminalId
    };
    self.sendRepMsg(backHeadNode);
};


TerminalControl.prototype.handle0072 = function (headNode, bodyNode) {
    logger.info('---handle0072---');
    logger.info(bodyNode.data);
};


TerminalControl.prototype.handle0032 = function (headNode, bodyNode) {
    var self = this;
    logger.info('---handle' + headNode.cmd + '---');
    //0032/0033用于处理错误票,此时应先将该票放入失败库,然后通知平台票出失败了
    //首先获取该票所在的库
    logger.info(headNode);
    logger.info(bodyNode);
    if (!bodyNode.ticketId) {
        return;
    }
    var ticketId = bodyNode.ticketId;

    mongoDBUtil.db.collection('terminal', {safe: true}, function (err, collection) {
        collection.findOne({'id': headNode.terminalId}, function (err, terminal) {
            if (terminal) {
                var name = 'waitTerminal' + terminal.terminalName;
                mongoDBUtil.db.collection(name, {safe: true}, function (err, _coll) {
                    _coll.findAndRemove({'id': ticketId}, [], function (err, _ticket) {
                        mongoDBUtil.db.collection('failTicket', {safe: true}, function (err, __coll) {
                            if (!_ticket) {
                                logger.info('不存在的票据 : ' + bodyNode.ticketId);
                            } else {
                                //通知平台失败
                                platform.sendFail(_ticket);
                                //入失败库
                                delete _ticket['_id'];
                                //新增失败信息,失败代码
                                _ticket.failInfo = bodyNode.retCode;
                                _ticket.failType = bodyNode.retDesc;
                                _ticket.terminalReturnTime = new Date().getTime();

                                __coll.insert(_ticket, function () {
                                });
                                if (headNode.cmd == '0033') {
                                    log.info('---发送0055---');
                                    var backHeadNode0005 = {
                                        cmd: '0055',
                                        sequenceId: headNode.sequenceId,
                                        terminalid: headNode.terminalId,
                                        retCode: prop.retCode.success.code,
                                        retDesc: prop.retCode.success.desc
                                    };
                                    var backBodyNode0005 = {
                                        ticketid: _ticket.id
                                    };
                                    self.send0053(backHeadNode0005, backBodyNode0005);
                                }
                            }
                        })
                    })
                })
            }
        });
    })
};


TerminalControl.prototype.handle0054 = function (headNode, bodyNode) {
    var self = this;
    logger.info('---handle' + headNode.cmd + '---');
    //0032/0033用于处理错误票,此时应先将该票放入失败库,然后通知平台票出失败了
    //首先获取该票所在的库
    logger.info(bodyNode.retCode);
    logger.info(bodyNode.retDesc);
    self.failHandler(headNode, bodyNode);
};


TerminalControl.prototype.handle4000 = function (headNode, bodyNode) {
    var self = this;
    logger.info('----handle4000----');
    logger.info(headNode);
    //终端机要死机,先从等待返回库中移走这些票,再改数组中状态,最后修改数据库的状态
    //需要消息回执
    var terminalId = headNode.terminalId;
    //step 1,从等待返回库中移走这些票
    for (var i = 0; i < terminalArray.length; i++) {
        var terminal = terminalArray[i];
        if (terminal.terminal.id == terminalId) {
            terminalArray.splice(i, 1);
        }
    }
    self.deadHandler(headNode);
};


TerminalControl.prototype.deadHandler = function (headNode) {
    var self = this;
    //打印模式死机
    mongoDBUtil.db.collection('terminal', {safe: true}, function (err, collection) {
        collection.findAndModify({id: headNode.terminalId}, {}, {$set: {status: prop.terminal.status.notAvailable}}, {new: true}, function (err, terminal) {
            //通知界面
            if (terminal) {
                var msg = JSON.stringify(terminal);
                self.io.emit('statusChange', msg);
                mongoDBUtil.db.collection('waitTerminal' + terminal.terminalName, {safe: true}, function (err, _coll) {
                    _coll.find({}).toArray(function (err, tickets) {
                        async.each(tickets, function (ticket, callback) {
                            mongoDBUtil.db.collection('waitTerminalTicket', {safe: true}, function (err, _col) {
                                delete ticket['_id'];
                                _col.insert(ticket, function () {
                                    //已添加到等待取走库
                                    callback(null);
                                })
                            })
                        }, function (err) {
                            var _backHeadNode = {
                                cmd: '4000',
                                sequenceId: headNode.sequenceId,
                                retCode: prop.retCode.success.code,
                                retDesc: prop.retCode.success.desc,
                                terminalId: headNode.terminalId
                            };
                            self.sendRepMsg(_backHeadNode);
                        })
                    });
                    _coll.remove({}, {safe: true}, function () {
                    })
                })
            } else {
                logger.info('找不到的终端机' + headNode.terminalId);
            }
        })
    })
};

TerminalControl.prototype.connClose = function (remoteAddress) {
    var self = this;
    //把状态修改为不可用,从数组中移除.
    for (var i = 0; i < terminalArray.length; i++) {
        var terminal = terminalArray[i];
        if (terminal.terminal.address == remoteAddress) {
            terminalArray[i].status = terminalCons.terminal.status.notAvailable;
            terminalArray.splice(i, 1);
            mongoDBUtil.db.collection('terminal', {
                safe: true
            }, function (err, collection) {
                collection.findAndModify({
                    'id': terminal.terminal.id
                }, {
                    'id': -1
                }, {
                    $set: {
                        status: terminalCons.terminal.status.notAvailable
                    }
                }, {
                    new: true
                }, function (err, result) {
                    var msg = JSON.stringify(result);
                    self.io.emit('statusChange', msg);
                    console.log('更改终端机' + result.id + '状态为' + result.status);
                })
            })
        }
    }
};


TerminalControl.prototype.insert2History = function (terminalId, bonusInfo, seq) {
    mongoDBUtil.db.collection('TerminalPrintSuccess', {
        safe: true
    }, function (err, collection) {
        collection.findAndRemove({
            'ticketSeq': seq
        }, [], function (err, result) {
            if (result) {
                var bonus = 0;
                delete result['_id'];
                result.bonusCount = 0;
                if (bonusInfo.indexOf('请到体育彩票管理中心领奖') != -1) {
                    var arr = bonusInfo.split('\r\n');
                    async.each(arr, function (item) {
                        if (item.indexOf('合计') != -1) {
                            bonus = parseInt(item.replace(/\D/g, ''));
                        }
                    }, function (err) {
                    });

                    result.bonusInfo = bonusInfo;
                    result.status = prop.ticket.status.bigBonus;
                    result.bonus = bonus;
                    result.bonusCount = 1;
                    collection.insert(result, function () {
                        logger.info(result.id + '已中大奖');
                    })
                } else {
                    if (bonusInfo.indexOf('恭喜中奖') != -1) {
                        result.bonusCount = 1;
                        var arr = bonusInfo.split('\r\n');
                        async.each(arr, function (item) {
                            if (item.indexOf('合计') != -1) {
                                bonus = parseInt(item.replace(/\D/g, ''));
                            }
                        }, function (err) {
                        })
                    }
                    result.bonusTime = new Date().getTime();
                    result.bonus = bonus;
                    result.bonusInfo = bonusInfo;
                    result.bonusTerminalId = terminalId;
                    mongoDBUtil.db.collection('History', {
                        safe: true
                    }, function (err, _coll) {
                        _coll.insert(result, function () {
                            logger.info(result.ticketSeq + '已入历史库');
                        })
                    })
                }
            } else {
                logger.info(err);
            }
        })
    });
};

TerminalControl.prototype.failHandler = function (headNode, bodyNode) {
    var self = this;
    var ticketId = bodyNode.ticketId;
    mongoDBUtil.db.collection('terminal', {safe: true}, function (err, collection) {
        collection.findOne({'id': headNode.terminalId}, function (err, terminal) {
            if (terminal) {
                mongoDBUtil.db.collection('waitTerminal' + terminal.terminalName, {safe: true}, function (err, _coll) {
                    _coll.findAndRemove({'id': ticketId}, [], function (err, ticket) {
                        if (ticket) {
                            mongoDBUtil.db.collection('failTicket', {safe: true}, function (err, __coll) {
                                logger.info(ticketId);
                                //通知平台失败
                                var body = {};
                                platform.sendFail(ticket);
                                delete ticket['_id'];
                                //新增失败信息,失败代码
                                ticket.failInfo = bodyNode.retCode;
                                ticket.failType = bodyNode.retDesc;
                                ticket.terminalReturnTime = new Date().getTime();
                                //入失败库
                                __coll.insert(ticket, function () {
                                });
                                //通知终端机失败
                                logger.info('---发送0055---');
                                var backHeadNode0005 = {
                                    cmd: '0055',
                                    sequenceId: headNode.sequenceId,
                                    terminalid: headNode.terminalId,
                                    retCode: prop.retCode.success.code,
                                    retDesc: prop.retCode.success.desc
                                };
                                var backBodyNode0005 = {
                                    ticketid: ticket.id
                                };
                                self.send0053(backHeadNode0005, backBodyNode0005);
                            })
                        }
                    })
                })
            }
        });
    })
};


/**
 * 2.1.7.1    出票请求消息(命令码:0002)
     请求消息: [服务器->终端机]
 * @param headNode
 * @param bodyNode
 *
 * packagenum    4    int    总的包个数
 * namelen    4    int    保存文件名长度
 * namedata        var_string( 不定长)    文件名的内容
 * datalen    4    int    数据内容长度
 * data        var_string(不定长)    数据内容
 *
 */
TerminalControl.prototype.send0002 = function (headNode, bodyNode) {
    var self = this;
    async.waterfall([
        function (cb) {
            /**
             * 第一步,计算总长度
             * 1.headNode长度为48不变
             * 2.bodyNode的长度为12+namedata+data的
             */
            var totalLength = 52;
            for (var i = 0; i < bodyNode.tickets.length; i++) {
                totalLength += bodyNode.tickets[i].namelen + bodyNode.tickets[i].datalen + 8;
            }
            cb(null, totalLength)
        },
        function (totalLength, cb) {
            //logger.info(totalLength);
            //先写头
            var head = headNode.cmd + headNode.sequenceId + headNode.terminalId;
            var result = new Buffer(head, msgParam.encoding);
            //logger.info(totalLength);
            var buf = new Buffer(totalLength);
            buf.writeInt32BE(totalLength - 4, 0);
            result.copy(buf, 4, 0, result.length);
            //写入总的包个数
            result = new Buffer(4);
            result.writeUInt32BE(bodyNode.tickets.length, 0);
            result.copy(buf, 48, 0, result.length);
            //写入保存文件名长度,注意从此处开始要循环写入了
            //记录位置
            cb(null, buf)
        },
        function (buf, cb) {
            var result;
            var index = 52;
            for (var i = 0; i < bodyNode.tickets.length; i++) {
                result = new Buffer(4);
                //写保存文件名长度
                result.writeInt32BE(bodyNode.tickets[i].namelen, 0);
                result.copy(buf, index, 0, result.length);
                index += 4;
                //写入保存文件名
                result = new Buffer(bodyNode.tickets[i].namedata);
                result.copy(buf, index, 0, result.length);
                index += bodyNode.tickets[i].namelen;
                //写入数据内容长度
                result = new Buffer(4);
                result.writeInt32BE(bodyNode.tickets[i].datalen, 0);
                result.copy(buf, index, 0, result.length);
                index += 4;
                //写入数据内容
                result = new Buffer(bodyNode.tickets[i].data);
                result.copy(buf, index, 0, result.length);
                index += bodyNode.tickets[i].datalen;
            }
            cb(null, buf);

        },
        function (buf, cb) {
            self.sendBuf(buf);
            log.info('---send0002已发送---');
            console.log(headNode);
            console.log(bodyNode);
        }
    ], function (err, result) {
        if (err) {
            console.log('err: ', err); // -> null
            console.log('result: ', result); // -> 16
        }
    })
};


/**
 * 2.1.7.1    打印票(命令码:0004)
     请求消息: [服务器->终端机]
 * @param headNode
 * @param bodyNode
 *
 *
 */

TerminalControl.prototype.send0004 = function (headNode, bodyNode) {
    var self = this;
    async.waterfall([
        function (cb) {
            /**
             * 第一步,计算总长度
             * 1.headNode长度为48不变
             * 2.bodyNode长度为 52 + dataLen + nameLen + 8 + 4
             */
            var totalLength = 48 + bodyNode.nameLen + bodyNode.dataLen + 8 + 4;
            cb(null, totalLength)
        },
        function (totalLength, cb) {
            //先写头
            var head = headNode.cmd + headNode.sequenceId + headNode.terminalId;
            var result = new Buffer(head);
            var buf = new Buffer(totalLength);
            buf.writeInt32BE(totalLength - 4, 0);
            result.copy(buf, 4, 0, result.length);
            //写入总的包个数
            result = new Buffer(4);
            result.writeUInt32BE(bodyNode.packageNum, 0);
            result.copy(buf, 48, 0, result.length);
            //写入保存文件名长度
            result = new Buffer(4);
            result.writeInt32BE(bodyNode.nameLen, 0);
            result.copy(buf, 52, 0, result.length);
            //写入保存文件名
            result = new Buffer(bodyNode.nameData);
            result.copy(buf, 56, 0, bodyNode.nameLen);

            //写入数据长度
            result = new Buffer(4);
            result.writeInt32BE(bodyNode.dataLen, 0);
            result.copy(buf, 56 + bodyNode.nameLen, 0, result.length);

            //写入数据,转换中文
            result = new Buffer(bodyNode.data, 'base64');
            result.copy(buf, 60 + bodyNode.nameLen, 0, bodyNode.dataLen);
            cb(null, buf);
        },
        function (buf, cb) {
            self.sendBuf(buf);
            log.info('---send0004已发送---');
            log.info(headNode);
            log.info(bodyNode);
        }
    ], function (err, result) {
        if (err) {
            log.info('err: ', err); // -> null
        }
    })
};



/**
 * 2.1.7.1    终端机模式切换(命令码:0009)
     请求消息: [服务器->终端机]
 * @param headNode
 * @param bodyNode
 *
 *
 */
TerminalControl.prototype.send0009 = function (headNode, bodyNode) {
    var self = this;
    /**
     * 先计算总长度
     * head : 4+4+32+8
     * body : 4
     */
    var totalLength = 4 + 4 + 32 + 8 + 4;
    var head = headNode.cmd + headNode.sequenceId + headNode.terminalId;
    var result = new Buffer(head, msgParam.encoding);

    var buf = new Buffer(totalLength);
    buf.writeInt32BE(totalLength - 4, 0);
    result.copy(buf, 4, 0, result.length);

    result = new Buffer(4);
    result.writeInt32BE(bodyNode.switchPrize, 0);
    result.copy(buf, 48, 0, result.length);
    self.sendBuf(buf);
    log.info('---send0009已发送---');
    log.info(headNode);
    log.info(bodyNode);
};


TerminalControl.prototype.send0053 = function (headNode, bodyNode) {
    var self = this;

    /**
     * 先计算总长度
     * head : 4+4+32+8
     * body : 32
     */

    var head = headNode.cmd + headNode.sequenceId + headNode.terminalid + headNode.retCode + headNode.retDesc;
    var result = new Buffer(head, prop.encoding);
    var totalLength = result.length + 32 + 4;

    var buf = new Buffer(totalLength);
    buf.writeInt32BE(totalLength - 4, 0);
    result.copy(buf, 4, 0, result.length);


    //写入ticketId
    result = new Buffer(bodyNode.ticketid, prop.encoding);
    result.copy(buf, totalLength - 32, 0, result.length);

    self.sendBuf(buf);
    log.info('---send0053已发送---');
    log.info(headNode);
    log.info(bodyNode);
};


module.exports = TerminalControl;
