// print_constants/NodePlatCons.js
/**
 * Created by w44 on 15-1-12.
 */


var target='dev';

var node_plat_https={
    'devOpt': {
        hostname: 'localhost',
        port: 9088,
        path: '/mcp-filter/main/interface.htm',
        method: 'POST',
        headers :{
            'Content-Type': 'application/x-www-form-urlencoded',
            'method': 'POST'
        }
    },
    'runOpt': {
        hostname: '182.254.129.17',
        port: 9090,
        path: '/mcp-filter/main/interface.htm',
        method: 'POST',
        headers :{
            'Content-Type': 'application/x-www-form-urlencoded',
            'method': 'POST'
        }
    }
};


if(target == 'dev')
{
    exports.node_plat_opt = node_plat_https.devOpt;
}
else if(target == 'run')
{
    exports.node_plat_opt = node_plat_https.runOpt;
}


var head={};
head.userId = 'C0002';
head.userType = "CHANNEL";
head.digestType = "md5";

var key='abcde';

var command={
    queryOrder: 'P01',
    queryTicket: 'P06',
    ticketResponse: 'P02',
    sendNumbers: 'P03',
    queryTerm: 'CQ01'
};

var ticket_size=50;

exports.ticket_size=ticket_size;
exports.head=head;
exports.key=key;
exports.command=command;
