// print_constants/NodePlatCons.js
/**
 * Created by w44 on 15-1-12.
 */


var target='dev';

var node_plat_https={
    'devOpt': {
        hostname: '218.30.107.19',
        port: 9088,
        path: '/mcp-filter/main/interface.htm',
        method: 'POST'
    },
    'runOpt': {
        hostname: '182.254.129.17',
        port: 9090,
        path: '/mcp-filter/main/interface.htm',
        method: 'POST'
    }
};


if(target == 'dev')
{
    exports.node_plat_opt = node_plat_https.devOpt;
}
else if(target == 'run')
{
    exports.node_plat_opt = node_plat_https.runOpt;
}


var head={};
head.userId = 'C0001';
head.userType = "CHANNEL";
head.digestType = "md5";
var key='123456';

var command={
    queryOrder: 'P01',
    queryTicket: 'P06',
    ticketResponse: 'P02',
    sendNumbers: 'P03',
    queryTerm: 'CQ01'
};


var ticket = {};
ticket.status = {
    wait: '1000',
    in: '1100',
    bigBonus:'1101',
    print: '2000',
    bonus: '1999',
    success:1300,
    fail:1500
};

var retCode = {
    success: {
        code: '0000',
        desc: 'success'
    },
    error: {
        code: '9999',
        desc: 'error'
    }
};


var gamCodeArray = ['T01', 'T02', 'T03', 'T04', 'T05', 'T06', 'T51', 'T53', 'T54', 'T55'];

var ticket_size=50;

var mode = 'test';

exports.mode=mode;
exports.ticket_size=ticket_size;
exports.head=head;
exports.key=key;
exports.retCode=retCode;
exports.ticket=ticket;
exports.command=command;
exports.gamCodeArray=gamCodeArray;
